cloudify_aws/ec2/resources/image.py

```diff
@@ -77,28 +83,87 @@ def create(self, params):
 
+        dry_run = params.get(DRY_RUN, False)
+        image_ids = params.get(IMAGE_IDS, [])
+        owners = params.get(OWNERS, [])
+        executable_users = params.get(EXECUTABLE_USERS, [])
+        filters = params.get(FILTERS, [])
+        if not any([dry_run, image_ids, owners, executable_users, filters]):
+            return
+        self.describe_image_filters = {
+            DRY_RUN: dry_run,
+            IMAGE_IDS: image_ids,
+            OWNERS: owners,
+            EXECUTABLE_USERS: executable_users,
+            FILTERS: filters
+        }
 
-def prepare_describe_image_filter(params, iface):
-    iface.describe_image_filters = {
-        DRY_RUN: params.get(DRY_RUN, False),
-        IMAGE_IDS: params.get(IMAGE_IDS, []),
-        OWNERS: params.get(OWNERS, []),
-        EXECUTABLE_USERS: params.get(EXECUTABLE_USERS, []),
-        FILTERS: params.get(FILTERS, [])
-    }
-    return iface
+
+def clean_params(params):
+    if isinstance(params, dict):
+        new_params = {}
+        for key, val in params.items():
+            if isinstance(val, dict) or isinstance(val, list):
+                val = clean_params(val)
+            if val:
+                new_params[key] = val
+        return new_params
+    if isinstance(params, list):
+        new_params = []
+        for val in params:
+            if isinstance(val, dict) or isinstance(val, list):
+                val = clean_params(val)
+            if val:
+                new_params.append(val)
+        return new_params
 
 
 @decorators.aws_resource(EC2Image, resource_type=RESOURCE_TYPE)
 def prepare(ctx, iface, resource_config, **_):
     """Prepares an AWS EC2 Image"""
     # Save the parameters
-    ctx.instance.runtime_properties['resource_config'] = resource_config
-    iface = \
-        prepare_describe_image_filter(
-            resource_config.copy(),
-            iface)
-    ami = iface.properties
-    utils.update_resource_id(ctx.instance, ami.get(IMAGE_ID))
+    if not iface.describe_image_filters:
+        return
+    if not ctx.instance.runtime_properties.get('use_external_resource'):
+        ctx.instance.runtime_properties['resource_config'] = resource_config
+        iface.prepare_describe_image_filter(resource_config.copy())
+        ami = iface.properties
+        utils.update_resource_id(ctx.instance, ami.get(IMAGE_ID))
+
+
+@decorators.aws_resource(EC2Image, resource_type=RESOURCE_TYPE)
+def create(ctx, iface, resource_config, **_):
+    """Create an AWS EC2 Image"""
+    if ctx.instance.runtime_properties.get('use_external_resource'):
+        # if use_external_resource there we are using an existing image
+        return
+
+    params = utils.clean_params(
+        dict() if not resource_config else resource_config.copy())
+    if 'InstanceId' not in params:
+        params['InstanceId'] = utils.find_resource_id_by_type(
+            ctx.instance, 'cloudify.nodes.aws.ec2.Instances')
+    params = clean_params(params)
```

Comment:
We should only need to clean params once. Maybe we should do that only here and skip it in line 147.